Libinput and fbdev improvements for unl0kr

kcxt opened the pull request:
Add a framebuffer quirk to force the display to refresh after every draw, this is needed for some quirky panels and DRM/fbdev emulation setups. It requires that the display buffer is set to the same size as the whole framebuffer so that the entire display can be flushed at once if needed, otherwise there is tearing.

Also use pthreads and a small circular queue to stop the libinput driver from dropping inputs. The LVGL tickrate can't always keep up and this ensures that inputs aren't dropped.

The keypad support needs to be tested.

Cc: @Johennes

kcxt committed:
display: fbdev: add force_refresh quirk

This is needed specifically for Qualcomm devices using DRM fbdev
emulation, it is probably useful for other platforms using fbdev
emulation.

kcxt committed:
indev: libinput: implement event queue

Libinput doesn't buffer events, so hook up a dedicated thread per input
device and a small circular queue to buffer input events. This fixes the
dropped input issue so you can type much faster. If there are too many
events, the oldest events in the queue will be dropped.

Johennes commented on indev/libinput.c:
```diff
@@ -250,6 +267,83 @@ void libinput_read(lv_indev_drv_t * indev_drv, lv_indev_data_t * data)
+{
+  libinput_lv_event_t *evt = &state->points[state->end];
+
+  if (++state->end == MAX_EVENTS)
```
Would you also have to do something to `state->start` here to prevent `end` overtaking `start` which would then read later events first when accessing `start`?

Johennes commented on indev/libinput.c:
```diff
@@ -259,37 +353,24 @@ void libinput_read(lv_indev_drv_t * indev_drv, lv_indev_data_t * data)
-    default:
-      break;
+  pthread_mutex_lock(&state->event_lock);
+  if (!event_pending(state)) {
```
I think you can remove this `if` because `get_event` does the same test and returns `NULL` if it fails which is then skipped by the `if (evt)` below and ends up in the same unlock and return behavior.

Johennes commented on indev/libinput.c:
```diff
@@ -410,39 +491,51 @@ static void reset_scanned_devices(void) {
     case LIBINPUT_EVENT_TOUCH_UP:
-      state->button = LV_INDEV_STATE_REL;
+      evt = new_event(state);
+      touch_event = libinput_event_get_touch_event(event);
```
I think this isn't actually used anymore because the line below is commented out?

Johennes reviewed (commented):
Pretty amazing stuff! Thanks @calebccff!

I've only given it a quick look this morning. Will try to do some testing against https://gitlab.com/cherrypicker/unl0kr/-/merge_requests/21 this week.

@kisvegabor, I know, lv_drivers is supposed to be deprecated in future. It'll be very helpful to have these changes merged though because https://gitlab.com/cherrypicker/unl0kr still depends on it.

We already started migrating to lvgl `master` in https://gitlab.com/cherrypicker/unl0kr/-/merge_requests/20 and we'll port the changes to the fbdev driver and the libinput driver, too, as part of this.

kisvegabor reviewed (commented):
Really nice, thank you!

We can merge it here, but yeah... it'd be great to add these updates to master as well. :slightly_smiling_face:

Johennes commented on indev/libinput.c:
```diff
@@ -259,37 +353,25 @@ void libinput_read(lv_indev_drv_t * indev_drv, lv_indev_data_t * data)
+    state->last_event = *evt;
+
+  data->point = evt->point;
+  data->state = evt->pressed;
```
I think this is missing `data->key = evt->key_val`?

Johennes commented on indev/libinput.c:
```diff
@@ -457,50 +554,51 @@ static void read_pointer(libinput_drv_state_t *state, struct libinput_event *eve
 static void read_keypad(libinput_drv_state_t *state, struct libinput_event *event) {
   struct libinput_event_keyboard *keyboard_event = NULL;
   enum libinput_event_type type = libinput_event_get_type(event);
+  libinput_lv_event_t *evt = NULL;
```
This doesn't appear to ever be set to something else than `NULL`. I think you're missing an `evt = new_event(state);`  inside the `case LIBINPUT_EVENT_KEYBOARD_KEY` below?

Johennes commented on indev/libinput.c:
```diff
@@ -419,30 +502,44 @@ static void read_pointer(libinput_drv_state_t *state, struct libinput_event *eve
+      /* ignore more than 2 fingers to avoid overflowing the buffer */
+      if (libinput_event_touch_get_slot(touch_event) > 1)
+        return;
+      evt = new_event(state);
```
I think something has went missing here. `evt` is only set here for `LIBINPUT_EVENT_TOUCH_UP` but it's written into in all the other cases below?

calebccff replied:
The other cases here fall through, I split this out as "common handling code", I'll rewrite it to be a bit clearer.

calebccff replied:
oh I missed the LIBINPUT_EVENT_POINTER_* cases too... nice catch :sweat_smile:

calebccff commented:
Totally forgot to push :sweat_smile: sorry for the wait. The touch/pointer handling was actually pretty broken... Anyways, hopefully this is good to go now!

Johennes commented:
Thanks for updating the PR @calebccff and sorry for the delay. I will take a look at this and test it within unl0kr but I won't manage before next week, unfortunately.

Johennes commented on indev/libinput.c:
```diff
@@ -186,6 +185,7 @@ bool libinput_set_file_state(libinput_drv_state_t *state, char* dev_name)
  */
 void libinput_init(void)
 {
+  memset(&default_state, 0, sizeof(libinput_drv_state_t));
```
Doesn't this undo `default_state = { .event_lock = PTHREAD_MUTEX_INITIALIZER, };`?

I'm also not seeing any call to `pthread_mutex_init` in the corresponding [unl0kr MR](https://gitlab.com/cherrypicker/unl0kr/-/merge_requests/21/diffs). Could it be that we're accidentally using `event_lock` from `default_state` in the dynamic states in unl0kr? They are cleared with `lv_memzero` so `event_lock` should always result in `0` which might be what `PTHREAD_MUTEX_INITIALIZER` returns in this case.

I think it would actually be good to call `pthread_mutex_init` inisde of `libinput_init_state` because the caller shouldn't have to know anything about the thread handling inside of the driver. You may just have to make sure that `libinput_init` doesn't initialize the lock in `default_state` twice by calling `libinput_init_state` but not sure if that would actually be a problem

Johennes replied:
Proposal for how to fix this: https://github.com/calebccff/lv_drivers/pull/2

Johennes commented on indev/libinput.c:
```diff
@@ -410,39 +493,55 @@ static void reset_scanned_devices(void) {
   lv_disp_drv_t *drv = lv_disp_get_default()->driver;
 
+  /* ignore more than 2 fingers as it will only confuse LVGL */
+  touch_event = libinput_event_get_touch_event(event);
```
This triggers a warning on the console when the event is one of the `LIBINPUT_EVENT_POINTER_*` types. So for mice and trackpads this prints on every cursor motion or key press.

Johennes replied:
Proposal how to fix this: https://github.com/calebccff/lv_drivers/pull/1

Johennes commented on indev/libinput.c:
```diff
@@ -410,39 +493,55 @@ static void reset_scanned_devices(void) {
-      state->most_recent_touch_point.x = LV_CLAMP(0, state->most_recent_touch_point.x, drv->hor_res - 1);
-      state->most_recent_touch_point.y = LV_CLAMP(0, state->most_recent_touch_point.y, drv->ver_res - 1);
+      evt->point.x += libinput_event_pointer_get_dx(pointer_event);
+      evt->point.y += libinput_event_pointer_get_dy(pointer_event);
```
Upon testing with a trackpad, it doesn't let me move the cursor from the top left corner of the screen and always jumps back. I think it's caused by this part here. We used to add the delta from `libinput_event_pointer_get_dx/y` to the most recently read coordinates before. But here `evt` is a blank slot that we've zero'ed out so we're always starting in the top left corner regardless of where the cursor moved before.

We could look back in the ring buffer and add the value of `point` from the previous event. However, we'd also have to account for overrunning the two pointers in the buffer again and I'm not sure what value we'd set for `point` in that case. What might be better instead is to store `dx` and `dy` here and then evaluate either `x` and `y` or `dx` and `dy` in `libinput_read_state` by making use of `last_event`. I think we'd also have to change that function to always update `last_event` though regardless of whether more events are pending or not.

Johennes replied:
Proposal how to fix this: https://github.com/calebccff/lv_drivers/pull/3

Johennes reviewed (changes requested):
Sorry again for the delay. I reviewed the code and tested it with https://gitlab.com/cherrypicker/unl0kr/-/merge_requests/21 in unl0kr.

Pointer devices don't currently work correctly because their delta motion logic needs special handling with the new event buffer.

There is possibly also a problem with the mutex. I didn't notice any issues related to that when testing but I think the mutex initialization is missing and might just work accidentally in unl0kr.

Johennes committed:
Only call libinput_event_get_touch_event on touch events

Johennes committed:
Initialise mutex in libinput_init_state

Johennes commented on indev/libinput.c:
```diff
@@ -211,6 +211,8 @@ void libinput_init_state(libinput_drv_state_t *state, char* path)
   state->fds[0].events = POLLIN;
   state->fds[0].revents = 0;
 
+  pthread_create(&state->worker_thread, NULL, libinput_poll_worker, state);
```
I included this in https://github.com/calebccff/lv_drivers/pull/2

Johennes committed:
Handle pointer click and motion events using delta logic

kcxt committed:
libinput: insert dummy events to handle fast OSK typing

When typing fast with thumbs, you might hit a key while your other thumb
is already pressing a key, however this confuses indev quite a lot.

For example, you'll press a key (say 'h') with your right thumb, then
before releasing your thumb you'll press the 'e' key with your left
thumb, and then finally release the 'h' key followed by the 'e' key.

This should result in "he" being typed, but usually results in an "e".
Indev assumes that a pointer release will always be for whatever the
last position was. This is reasonable for mice but not for multitouch
displays.

This workaround detects this scenario and inserts two dummy events,
first a "pressed" state with the coordinates of the first finger (the
'h' key), then a release event, triggering the keypress, then it sends
another pressed events with the position of the second finger (the 'e'
key) so that if the next event is that finger releasing, the position
will be correct.

calebccff commented:
Rebased on master and picked up Johannes' fixes (Thanks a lot for that!). Should be good to go! Tested again with unl0kr.

I added a new commit to properly handle typing with two thumbs by injecting some fake events, with this you can type as fast as you want with your thumbs and not miss an input which is really nice for unl0kr. See the commit description for more details.

I get that this is a bit of a hack, I'm happy to drop it if it blocks the rest of the series.

Johennes commented on indev/libinput.c:
```diff
@@ -410,50 +550,79 @@ static void reset_scanned_devices(void) {
+      evt->point.x += libinput_event_pointer_get_dx(pointer_event);
+      evt->point.y += libinput_event_pointer_get_dy(pointer_event);
+      evt->point.x = LV_CLAMP(0, evt->point.x, drv->hor_res - 1);
+      evt->point.y = LV_CLAMP(0, evt->point.y, drv->ver_res - 1);
```
I think something may have gone wrong in the cherry picking. This part looks different from https://github.com/calebccff/lv_drivers/pull/3/files. The clamping and the use of `+=` shouldn't be here anymore as they've moved to `libinput_read_state` above. When testing the current branch with unl0kr and a track pad, I can only move the cursor down, not up anymore and I think this is the reason.

Johennes replied:
Proposal for fix: https://github.com/calebccff/lv_drivers/pull/4

Johennes commented on indev/libinput.c:
```diff
@@ -259,37 +365,71 @@ void libinput_read(lv_indev_drv_t * indev_drv, lv_indev_data_t * data)
+    /* Finally, update the position to the remaining finger and send a press */
+    evt->pressed = LV_INDEV_STATE_PR;
+    evt->point = state->slots[1].point; /* Wherever slot 1 most recently pressed */
+    state->doing_mtouch_dummy_event = 0;
```
I think you can consume the event here? If I'm reading the code correctly, it'll otherwise be hit again on the next call of this function in which it should enter the `else` right below. It's probably a no-op because it's just triggering the same press event twice but probably worth avoiding.

calebccff replied:
Nice catch! I think you're right.

Johennes commented:
> Rebased on master and picked up Johannes' fixes (Thanks a lot for that!). Should be good to go! Tested again with unl0kr.
> 
> I added a new commit to properly handle typing with two thumbs by injecting some fake events, with this you can type as fast as you want with your thumbs and not miss an input which is really nice for unl0kr. See the commit description for more details.
> 
> I get that this is a bit of a hack, I'm happy to drop it if it blocks the rest of the series.

The `slot` stuff is clever!

One possible issue I'm seeing is that you're writing `state->slots` from both the polling thread and the LVGL input handler. The locking is avoiding concurrent writes. But depending on how fast events enter and leave the queue, I think it's possible for later multi-touch events to already have overridden the `slots` array filled by earlier ones before you could use it?

I guess generally I'd be nice if LVGL itself supported multi-touch but the event injection strikes me as an acceptable workaround until then (especially given that we already do some event rewriting with `is_relative`).

Johennes commented on indev/libinput.c:
```diff
@@ -410,50 +550,79 @@ static void reset_scanned_devices(void) {
+      evt->point.y += libinput_event_pointer_get_dy(pointer_event);
+      evt->point.x = LV_CLAMP(0, evt->point.x, drv->hor_res - 1);
+      evt->point.y = LV_CLAMP(0, evt->point.y, drv->ver_res - 1);
+      evt->is_relative = true;
```
One more issue that I just found is that the code currently doesn't support dragging anymore. E.g. if you click and hold but even just slightly move the cursor, that triggers a release because we're not setting `evt->pressed` here.

Unfortunately, `libinput_event_pointer_get_button_state(..)` doesn't seem to work for motion events. So we cannot use it here.

We could maybe define two relativity flags, `is_relative_position` (which would mean add up `x` and `y` to the latest state) and `is_relative_button` (which would mean inherit the `pressed` value from the latest state). `LIBINPUT_EVENT_POINTER_MOTION` events would then require both `is_relative_position` and `is_relative_button` while  `LIBINPUT_EVENT_POINTER_BUTTON` events would only require `is_relative_position`.

Does that make sense?

calebccff replied:
I haven't tested it but I just added a `pointer_button_down` property to `libinput_drv_state_t`, had that attribute be updated on `LIBINPUT_EVENT_POINTER_BUTTON` and set `evt->pressed` based on that. I also made `LIBINPUT_EVENT_POINTER_BUTTON` fallthrough to `LIBINPUT_EVENT_POINTER_MOTION` so that `evt->point` is set.

I think that should work?

Johennes replied:
Nice! Yes, that seems to work. This now actually makes me think we could do the same for the delta movements. Instead of doing the weird `is_relative` thing when consuming events, just store last position we've seen in `state` and add the delta from libinput to it.

Johennes replied:
Have made it so.

Johennes commented:
> One possible issue I'm seeing is that you're writing `state->slots` from both the polling thread and the LVGL input handler. The locking is avoiding concurrent writes. But depending on how fast events enter and leave the queue, I think it's possible for later multi-touch events to already have overridden the `slots` array filled by earlier ones before you could use it?

Actually, I think you can work around that by doing the event injection when reading from libinput as opposed to when consuming the cached events. The `slots` array is guaranteed to only be filled by events preceding the current one at that point.

kcxt committed:
libinput: fix pointer always releasing on motion event

Johennes committed:
Eliminate is_relative in favor of storing last known position

Johennes committed:
Also set pressed on absolute pointer motion events

Johennes committed:
Move dummy event logic for multi-touch into the libinput read loop to make it thread-safe

Johennes committed:
Handle the case where the second finger is released before the first

Johennes commented:
> Actually, I think you can work around that by doing the event injection when reading from libinput as opposed to when consuming the cached events. The `slots` array is guaranteed to only be filled by events preceding the current one at that point.

This has been implemented, too, now.

Johennes commented:
@kisvegabor this is now done from our end and ready for your review. I've tested it correctly in unl0kr with touch, pointer and keyboard devices. Once we make another unl0kr release, I'll start working on porting this into lvgl `master`.

Johennes committed:
Set point on touch release events

jane400 committed:
libinput: release keyboard keys in lvgl directly after pressing

This enables fast typing on hardware keyboards as keys are no longer forgotten.

Johennes committed:
Merge pull request #5 from jane400/jane/unl0kr-keyboard-hack

hack for hardware keyboard

kisvegabor commented:
Sorry, I've just noticed this PR when the stale bot started to work again after years...

Is it still relevant?
